### PR TITLE
KAFKA-10149: Allow auto preferred leader election when there are ongoing partition reassignments

### PR DESCRIPTION
**Pull request by songnon**
Auto preferred leader election should be allowed even if there are ongoing reassignments. Otherwise, it might result in unbalanced leaders if there are any long-running reassignments. For partitions being reassigned, a leader election will only be triggered when the TRS = ISR as part of the partition reassignment procedure automatically if the current leader is not in TRS or isn't alive (step [B3](https://github.com/apache/kafka/blob/f83d95d9a28267f7ef7a7b1e584dcdb4aa842210/core/src/main/scala/kafka/controller/KafkaController.scala#L699)).

For example, let's say there are broker `(1,2,3,4)`, and a partition is current on `(1,2)` , 1 is the current leader:
- if we reassign that partition to `(3, 4)`, a leader election would be triggered during the reassignment step B3, and 3 will be the new leader.
- if we reassign that partition to `(3, 1)`, leader election **won't** be triggered during the reassignment as the current leader is still in the TRS, 1 will continue to be the leader. The auto preferred leader election will help in such situation.
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala [hunk 1]
```diff
@@ -490,6 +490,41 @@ class ControllerIntegrationTest extends QuorumTestHarness {
       "failed to get expected partition state upon broker startup")
   }
 
+  @Test
+  def testAutoPreferredReplicaLeaderElectionWithReassigningPartitions(): Unit = {
```
Comment (jolshan) [inline core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: I think it would still work fine, but did we consider a situation where the preferred leader election occurs right before or at the same time as the reassignment completes? 
For example, we have a reassignment going, a broker is restarted, and we may or may not have preferred leader election before it completes.

File: core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala [hunk 2]
```diff
@@ -490,6 +490,41 @@ class ControllerIntegrationTest extends QuorumTestHarness {
...
+    zkClient.createPartitionReassignment(reassignment.map { case (k, v) => k -> v.replicas })
+    waitForPartitionState(reassigningTp, firstControllerEpoch, controllerId, LeaderAndIsr.InitialLeaderEpoch + 1,
+      "failed to get expected partition state during partition reassignment with offline replica")
+
+    servers(leaderBrokerId).startup()
+    waitForPartitionState(tp, firstControllerEpoch, leaderBrokerId, LeaderAndIsr.InitialLeaderEpoch + 2,
+      "failed to get expected partition state upon broker startup")
+
+    servers(otherBrokerId).startup()
+    waitForPartitionState(reassigningTp, firstControllerEpoch, otherBrokerId, LeaderAndIsr.InitialLeaderEpoch + 4,
```
Comment (jolshan) [inline core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala:532]: Is this leader epoch +4 because the startup and the reassignment?

Comment (songnon) [reply core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala:532]: Yes, it is.

File: core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala (hunk 1 above)
Comment (songnon) [reply core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: During a partition's reassignment, if the current leader is not in TRS or isn't alive, a leader election will be triggered at the time when the TRS = ISR. At that time, the partition is still in `controllerContext.partitionsBeingReassigned`, so `checkAndTriggerAutoLeaderRebalance` will skip that particular partition. 
When the reassignment is done (partition being removed from `controllerContext.partitionsBeingReassigned`),  `checkAndTriggerAutoLeaderRebalance` will treat it as a normal partition, preferred leader election may be triggered if it's allowed to.

File: core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala (hunk 2 above)
Comment (jolshan) [reply core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala:532]: Got it -- wasn't sure if it was worth adding a partition state for epoch + 3 or leaving a comment about what we expect.

Comment (songnon) [reply core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala:532]: I guess the scenario for partition `reassigningTp` in this test case is a very typical scenario. We have a similar test `testPartitionReassignmentResumesAfterReplicaComesOnline` in this file. I just use it to simulate an ongoing reassignment. I can add some comments if it's necessary.

**Comment by songnon** [comment]
Update the code and commit message. It looks like partitions being reassigned can also benefit from auto preferred leader election

File: core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala [hunk 3]
```diff
@@ -490,6 +490,41 @@ class ControllerIntegrationTest extends QuorumTestHarness {
...
+    servers(leaderBrokerId).shutdown()
+    servers(leaderBrokerId).awaitShutdown()
+
+    servers(otherBrokerId).shutdown()
+    servers(otherBrokerId).awaitShutdown()
+    waitForPartitionState(tp, firstControllerEpoch, controllerId, LeaderAndIsr.InitialLeaderEpoch + 1,
+      "failed to get expected partition state upon broker shutdown")
+
+    val reassignment = Map(reassigningTp -> ReplicaAssignment(Seq(otherBrokerId), List(), List()))
+    zkClient.createPartitionReassignment(reassignment.map { case (k, v) => k -> v.replicas })
```
Comment (C0urante) [inline core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala:520]: Can this be simplified? Don't think we need to construct an entire `ReplicaAssignment` here:
```suggestion
    val reassignment = Map(reassigningTp -> Seq(otherBrokerId))
    zkClient.createPartitionReassignment(reassignment)
```

**Review by C0urante — COMMENTED** [review]
Should we also add a test case for when the leader election takes place on the same topic as the one undergoing reassignment?

File: core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala (hunk 3 above)
Comment (songnon) [reply core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala:520]: You are right. Thanks.

**Comment by songnon** [comment]
> Should we also add a test case for when the leader election takes place on the same topic as the one undergoing reassignment?

@C0urante Added a new test case. Could you have a look? Thanks.

File: core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala [hunk 4]
```diff
@@ -490,6 +490,58 @@ class ControllerIntegrationTest extends QuorumTestHarness {
...
+    val otherBrokerId = servers.map(_.config.brokerId).filter(e => e != controllerId && e != leaderBrokerId).head
+    val tp = new TopicPartition("t", 0)
+    val assignment = Map(tp.partition -> Seq(leaderBrokerId, controllerId))
+    TestUtils.createTopic(zkClient, tp.topic, partitionReplicaAssignment = assignment, servers = servers)
+    val reassignment = Map(tp -> Seq(otherBrokerId, leaderBrokerId))
+    zkClient.createPartitionReassignment(reassignment)
+    waitForPartitionState(tp, firstControllerEpoch, leaderBrokerId, LeaderAndIsr.InitialLeaderEpoch + 3,
+      "failed to get expected partition state after partition reassignment")
+    waitForPartitionState(tp, firstControllerEpoch, otherBrokerId, LeaderAndIsr.InitialLeaderEpoch + 4,
+      "failed to get expected partition state after auto preferred replica leader election")
```
Comment (C0urante) [inline core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: Is it possible that the reassignment completes before the next automatic leader election is triggered, in which case we don't actually verify that the election has taken place while the reassignment is active?

Comment (songnon) [reply core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala]: hmm, that's a good point. Updated the test case to make sure election is taken place while the reassignment is active.

File: core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala (hunk 3 above)
Comment (songnon) [reply core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala:520]: Enhanced the test case a bit. Now it's required to construct an entire `ReplicaAssignment`

**Review by C0urante — APPROVED** [review]
Thanks @songnon, LGTM. Test failures appear unrelated.

@jolshan do you have any other thoughts? Also, FYI it looks like there was a failure in the newly-introduced `ProducerIdExpirationTest` in the run [here](https://ci-builds.apache.org/blue/rest/organizations/jenkins/pipelines/Kafka/pipelines/kafka-pr/branches/PR-12543/runs/8/nodes/13/steps/114/log/?start=0), with the `testTransactionAfterTransactionIdExpiresButProducerIdRemains` case.

**Comment by jolshan** [comment]
@C0urante thanks for reminding me about this. I think the last thing I want to confirm is that in the case where we have a reassignment in progress -- we don't do the preferred leader election. I see we added the test case, but its a bit hard to parse. 

I'm also a little confused by shutting down the broker that is not part of the assignment in the beginning. Some comments there could really help me understand the scenario better.

Looks like the producer ID test didn't fail again, so it might be flaky. I will keep an eye on it. Looks like the builds are also red now, but maybe that's an issue on master.

**Comment by C0urante** [comment]
Thanks @jolshan. I believe it's valid to perform preferred leader election for topics even if they're currently undergoing reassignment, for reasons [discussed in an earlier PR](https://github.com/apache/kafka/pull/9302#discussion_r491094781) for this ticket by @hachikuji:
> During a reassignment, the adding replicas are always listed first which means the preferred leader is among the target replicas. My take is that we want to move the leadership onto the new preferred leader as soon as possible since the whole point of the reassignment is to take load off the removing replicas.

I've double-checked the code base and this still appears to be the case; let me know if I'm missing something, though.

Definitely agree RE commenting the test code; @songnon if you have time would you mind adding some details? Explanation of why we're shutting down and restarting which brokers when (e.g., to create a stuck reassignment, or to create a pending preferred leader election) should be plenty.

**Comment by jolshan** [comment]
> I believe it's valid to perform preferred leader election for topics even if they're currently undergoing reassignment

Got it. I think the PR description here confused me then.
> Auto preferred leader election should be allowed for partitions which don't have any ongoing reassignments.

Perhaps we can update this to make it clearer that we can also do the preferred leadership election on partitions undergoing reassignment to update the preferred leader sooner.

**Comment by songnon** [comment]
> Perhaps we can update this to make it clearer that we can also do the preferred leadership election on partitions undergoing reassignment to update the preferred leader sooner.

Fixed. I forgot to update the PR description after making some code change and updating the commit text.

**Comment by songnon** [comment]
> I'm also a little confused by shutting down the broker that is not part of the assignment in the beginning. Some comments there could really help me understand the scenario better.

@jolshan @C0urante I have added some comments in the test code. Hope it helps. Let me know if there are any questions.

**Review by C0urante — APPROVED** [review]
Thank you for your continued diligent work on this fix, @songnon. LGTM!

**Comment by ijuma** [comment]
@junrao Was there an underlying reason not to allow this previously?

**Comment by junrao** [comment]
Previously, the controller logic wasn't single threaded. So, this check is probably just to prevent concurrent actions. Now that all logic in the controller runs in a single thread, this constraint could be revisited.
